### PR TITLE
More e2e tests and the beginings of a script that could be called by the new pipeline

### PR DESCRIPTION
**Pull request by asalkeld**
The script works nicely in a dev environment, I am still unsure of what makes sense from the pipelines perspective.

### CODE REVIEW

File: hack/e2e.sh [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        done
+        az network vnet subnet update \
+            -g "$RESOURCEGROUP" \
+            --vnet-name dev-vnet \
+            -n "$CLUSTER-master" \
+            --disable-private-link-service-network-policies true \
+            >/dev/null
+
+        echo "====== aro create =========================="
+        az aro create -g "$RESOURCEGROUP" --cluster-resource-group "$CLUSTER" -n "$CLUSTER" --vnet dev-vnet --master-subnet "$CLUSTER-master" --worker-subnet "$CLUSTER-worker"
```
Comment (asalkeld) [inline hack/e2e.sh]: I had some trouble using --cluster-resource-group, will confirm if it works (I wanted to distinguish the e2e clusters from dev ones)

File: hack/e2e.sh [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    ./aro rp &
+    trap 'return_id=$?; aro_ci_teardown_handler; exit $return_id' EXIT
+    while [[ "$(curl -s -o /dev/null -w '%{http_code}' localhost:8443)" == "000" ]]; do
+        sleep 2
+    done
+
+    if $CLUSTER_CREATE; then
+        echo "====== setup subnets =========================="
+        for subnet in "$CLUSTER-master" "$CLUSTER-worker"; do
+            az network vnet subnet create \
```
Comment (mjudeikis) [inline hack/e2e.sh]: Should we just add those into `az` as helpers and eliminate them from here? @jim-minter ?

**Comment by mjudeikis** [comment]
overall looks nice :)

File: hack/e2e.sh [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    index=$(($RANDOM % $size))
+    echo ${array[$index]}
+}
+
+aro_ci_setup() {
+    echo "====== starting dev RP =========================="
+    go build ./cmd/aro
+    ./aro rp &
+    trap 'return_id=$?; aro_ci_teardown_handler; exit $return_id' EXIT
+    while [[ "$(curl -s -o /dev/null -w '%{http_code}' localhost:8443)" == "000" ]]; do
```
Comment (jim-minter) [inline hack/e2e.sh]: poll on https://localhost:8443/healthz/ready returns 200

File: hack/e2e.sh (hunk 2 above)
Comment (jim-minter) [reply hack/e2e.sh]: I think this is a valid architectural topic to bring up with MSFT - please could you?  My current view is that on balance I don't think it should be implemented in `az`, but it's not a super strong feeling and I'm open to be persuaded.

File: hack/e2e.sh (hunk 1 above)
Comment (jim-minter) [reply hack/e2e.sh]: what trouble? yes, please check.

File: hack/e2e.sh [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            az network vnet subnet delete \
+                -g "$RESOURCEGROUP" \
+                --vnet-name dev-vnet \
+                -n "$subnet" \
+                >/dev/null
+        done
+    fi
+    echo "====== kill the RP =========================="
+    kill $(lsof -t -i :8443)
+    wait $(lsof -t -i :8443)
```
Comment (jim-minter) [inline hack/e2e.sh]: race condition - do the lsof once before the kill

File: hack/e2e.sh [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+get_random_location() {
+    array[0]="australiasoutheast"
+    array[1]="eastus"
+    array[2]="westeurope"
+
+    size=${#array[@]}
+    index=$(($RANDOM % $size))
+    echo ${array[$index]}
+}
```
Comment (jim-minter) [inline hack/e2e.sh]: can leave this for now but I'm not convinced - a) hard codes regions in the script, b) this kind of thing has caused us problems in v3

File: hack/e2e.sh [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    export CLUSTER=v4-e2e-$(git log --format=%h -n 1 HEAD)
+fi
+
+export RESOURCEGROUP="v4-$LOCATION"
+export COSMOSDB_ACCOUNT="$RESOURCEGROUP"
+export DOMAIN_NAME="$RESOURCEGROUP"
+export KEYVAULT_NAME="$RESOURCEGROUP"
+export PROXY_HOSTNAME="vm0.aroproxy.$LOCATION.cloudapp.azure.com"
+export DATABASE_NAME="e2e-$(git log --format=%h -n 1 HEAD)"
+export KUBECONFIG=$(pwd)/$CLUSTER.kubeconfig
```
Comment (jim-minter) [inline hack/e2e.sh]: can trim the unused ones of these.  Presumably all this ends up in the configuration secret?

File: pkg/util/azureclient/mgmt/redhatopenshift/operations.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: true,
+				},
+			},
+		}
+	} else {
+		client = redhatopenshift.NewOperationsClient(subscriptionID)
+		client.Authorizer = authorizer
+	}
+	client.PollingDuration = time.Minute * 60
```
Comment (jim-minter) [inline pkg/util/azureclient/mgmt/redhatopenshift/operations.go]: not needed here

File: test/e2e/checknodes.go [hunk 8]
```diff
@@ -29,8 +29,7 @@ var _ = Describe("Check nodes", func() {
 
 		nodes, err := Clients.Kubernetes.CoreV1().Nodes().List(metav1.ListOptions{})
 		Expect(err).NotTo(HaveOccurred())
-
-		var nodeCount int64
+		var nodeCount int32
```
Comment (jim-minter) [inline test/e2e/checknodes.go]: fyi you can also say `var nodeCount int` and later `Expect(nodeCount).To(BeEquivalentTo(expectedNodeCount))`

https://onsi.github.io/gomega/#beequivalenttoexpected-interface

File: test/e2e/operations.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+)
+
+var _ = Describe("List operations", func() {
+	Specify("the correct static operations are returned", func() {
+		ctx := context.Background()
+
+		opList, err := Clients.Operations.List(ctx)
+		Expect(err).NotTo(HaveOccurred())
+
+		l := &redhatopenshift.OperationList{
```
Comment (jim-minter) [inline test/e2e/operations.go]: feels a bit brittle, I think testing len(l.Operations) > 0 would probably be ok.  If we want to test the detail a unit test would make sense.

File: hack/e2e.sh [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    wait $(lsof -t -i :8443)
+}
+
+# allow overriding these variables
+if [[ -z "$CLUSTER_CREATE" ]]; then
+    export CLUSTER_CREATE=true
+fi
+if [[ -z "$CLUSTER_DELETE" ]]; then
+    export CLUSTER_DELETE=true
+fi
```
Comment (jim-minter) [inline hack/e2e.sh]: is this just for testing this script? is it needed?

Comment (asalkeld) [reply hack/e2e.sh]: i can remove it (yeah mostly for my testing)

**Comment by jim-minter** [comment]
looks good - OK for me to merge @asalkeld?

**Pull request merged**